README.md
# Go programming language plugin for Jenkins

[![Jenkins plugin](https://img.shields.io/jenkins/plugin/v/golang.svg)](https://plugins.jenkins.io/golang)
[![Jenkins plugin installs](https://img.shields.io/jenkins/plugin/i/golang?color=blue)](https://plugins.jenkins.io/golang)
[![Build status](https://ci.jenkins.io/job/Plugins/job/golang-plugin/job/master/badge/icon)](https://ci.jenkins.io/job/Plugins/job/golang-plugin/job/master/)

Automatically installs and sets up the [Go programming language](http://golang.org/) (golang) tools on a Jenkins agent during a build.

## Functionality
During a build, this plugin can:
- Install a particular version of Go on the agent that the build is running on
  - The correct package for the machine's operating system and CPU architecture will be automatically downloaded and installed, if not already present
- Export the `GOROOT` environment variable, pointing to the installed Go tools
- Add the path `$GOROOT/bin` to the `PATH`, so that the tools are available during the build

## Usage
### Global configuration
1. In the Jenkins global tool configuration settings (Manage Jenkins → Global Tool Configuration), find the "Go" section, click "Go Installations…" and "Add Go".
2. Enter a name, e.g. "Go 1.15" — the name itself has no significance, but will be displayed to users during Freestyle job configuration, or is what you need to enter as the `name` in a Pipeline
3. Either select "Install automatically" and select the desired Go version from the drop-down list or specify the installation directory manually

### Per-job configuration
#### Freestyle
1. In a job's configuration, find the "Build environment" section
2. Select the "Set up Go programming language tools" checkbox
3. Select the name of a Go installation from the drop-down

#### Pipeline
As with any other type of Tool Installer, you can use the `tool` step, in this case with the `go` tool type.

For example, with a Scripted Pipeline:

```groovy
// Run on an agent where we want to use Go
node {
    // Ensure the desired Go version is installed
    def root = tool type: 'go', name: 'Go 1.15'

    // Export environment variables pointing to the directory where Go was installed
    withEnv(["GOROOT=${root}", "PATH+GO=${root}/bin"]) {
        sh 'go version'
    }
}
```

## Changelog
See [CHANGELOG.md](https://github.com/jenkinsci/golang-plugin/blob/master/CHANGELOG.md).


CHANGELOG.md
# Version history

## Version 1.4
January 27, 2021

- Added support for installing Go on machines with 32-bit ARM CPU architecture

## Version 1.3
December 20, 2020

- Added support for installing Go on machines with 64-bit ARM (AArch64) CPU architecture ([#3](https://github.com/jenkinsci/golang-plugin/pull/3); thanks to [Sergei](https://github.com/serges147))
- Increased required Jenkins version to 2.190.3

## Version 1.2
February 25, 2017

- Stopped breaking tool configuration page if the version metadata had not been fetched (see [JENKINS-27499](https://issues.jenkins-ci.org/browse/JENKINS-27499))
- Added `go` symbol, for use with the Pipeline `tool` step (see [JENKINS-34345](https://issues.jenkins-ci.org/browse/JENKINS-34345))
- Increased required Jenkins version to 1.642.3

## Version 1.1
June 21, 2014

- Worked around bug causing newer versions of Go to not install (see [JENKINS-23509](https://issues.jenkins-ci.org/browse/JENKINS-23509))
- Fixed bug where wrong package for OS X could be installed (see [JENKINS-23505](https://issues.jenkins-ci.org/browse/JENKINS-23505))
- Ensured "Install automatically" is checked by default when adding a Go installation

## Version 1.0
June 18, 2014

-   Initial release


src/test/java/org/jenkinsci/plugins/golang/GolangInstallerTest.java
package org.jenkinsci.plugins.golang;

import org.junit.Test;

import javax.annotation.Nullable;

import static org.jenkinsci.plugins.golang.GolangInstaller.GolangInstallable;
import static org.jenkinsci.plugins.golang.GolangInstaller.GolangRelease;
import static org.jenkinsci.plugins.golang.GolangInstaller.InstallationFailedException;
import static org.junit.Assert.assertEquals;

public class GolangInstallerTest {

    // These are definitions of available Go packages, as would be listed in the Jenkins installer JSON
    private static final GolangInstallable FREEBSD_32 = createPackage("freebsd", "386");
    private static final GolangInstallable FREEBSD_64 = createPackage("freebsd", "amd64");
    private static final GolangInstallable LINUX_32 = createPackage("linux", "386");
    private static final GolangInstallable LINUX_64 = createPackage("linux", "amd64");
    private static final GolangInstallable LINUX_ARM32 = createPackage("linux", "armv6l");
    private static final GolangInstallable LINUX_ARM64 = createPackage("linux", "arm64");
    private static final GolangInstallable OS_X_10_6_32 = createPackage("darwin", "386", "10.6");
    private static final GolangInstallable OS_X_10_6_64 = createPackage("darwin", "amd64", "10.6");
    private static final GolangInstallable OS_X_10_8_32 = createPackage("darwin", "386", "10.8");
    private static final GolangInstallable OS_X_10_8_64 = createPackage("darwin", "amd64", "10.8");
    // As of Go 1.5, there is only a single OS X build distributed, with no OS X version set
    private static final GolangInstallable OS_X_GO_1_5 = createPackage("darwin", "amd64");

    @Test(expected = InstallationFailedException.class)
    public void testUnsupportedOs() throws InstallationFailedException {
        // Given we have configured a release we want to install
        GolangRelease release = createReleaseInfo();

        // When we try to get the install package for an OS which is not supported
        GolangInstaller.getInstallCandidate(release, "Android", "armv7a", null);

        // Then an exception should be thrown
    }

    @Test
    public void testFreeBsdInstallation() throws InstallationFailedException {
        // Given we have configured a release we want to install
        GolangRelease release = createReleaseInfo();

        // When we try to get the install package for FreeBSD
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "FreeBSD", "amd64", "10.2-RELEASE");

        // Then we should get the correct FreeBSD package
        assertEquals("Got unexpected package", FREEBSD_64, pkg);
    }

    @Test
    public void testFreeBsd32BitInstallation() throws InstallationFailedException {
        // Given we have configured a release we want to install
        GolangRelease release = createReleaseInfo();

        // When we try to get the install package for FreeBSD
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "FreeBSD", "i386", "10.2-RELEASE");

        // Then we should get the correct FreeBSD package
        assertEquals("Got unexpected package", FREEBSD_32, pkg);
    }

    @Test(expected = InstallationFailedException.class)
    public void testUnsupportedFreeBsd32BitInstallation() throws InstallationFailedException {
        // Given we have configured a Go 1.5 release we want to install
        GolangRelease release = createReleaseInfo(OS_X_GO_1_5, FREEBSD_64, LINUX_32, LINUX_64, LINUX_ARM64);

        // When we try to get the install package for 32-bit FreeBSD
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "FreeBSD", "i386", "10.2-RELEASE");

        // Then an exception should be thrown, as there is no 32-bit package
    }

    @Test
    public void testLinuxArm32BitInstallation() throws InstallationFailedException {
        // Given we have configured a release we want to install
        GolangRelease release = createReleaseInfo();

        // When we try to get the install package for Linux ARM 32-bit (aka `arm`)
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "linux", "arm", "5.4.0");

        // Then we should get the correct Linux ARM 32-bit package
        assertEquals("Got unexpected package", LINUX_ARM32, pkg);
    }

    @Test
    public void testLinuxArm32BitInstallationAlias() throws InstallationFailedException {
        // Given we have configured a release we want to install
        GolangRelease release = createReleaseInfo();

        // When we try to get the install package for Linux ARM 32-bit, with the CPU value `aarch32`
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "linux", "aarch32", "5.4.0");

        // Then we should get the correct Linux ARM 32-bit package
        assertEquals("Got unexpected package", LINUX_ARM32, pkg);
    }

    @Test
    public void testLinuxAarch64BitInstallation() throws InstallationFailedException {
        // Given we have configured a release we want to install
        GolangRelease release = createReleaseInfo();

        // When we try to get the install package for Linux ARM 64-bit (aka `aarch64`)
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "linux", "aarch64", "5.4.0");

        // Then we should get the correct Linux ARM 64-bit package
        assertEquals("Got unexpected package", LINUX_ARM64, pkg);
    }

    @Test
    public void testLinuxArm64BitInstallationAlias() throws InstallationFailedException {
        // Given we have configured a release we want to install
        GolangRelease release = createReleaseInfo();

        // When we try to get the install package for Linux ARM 64-bit, with the CPU value `arm64`
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "linux", "arm64", "5.4.0");

        // Then we should get the correct Linux ARM 64-bit package
        assertEquals("Got unexpected package", LINUX_ARM64, pkg);
    }

    @Test
    public void testLatestGo15PackageForOsXVersionReturned() throws InstallationFailedException {
        // Given we have configured a Go 1.5 release we want to install
        GolangRelease release = createReleaseInfo(LINUX_32, LINUX_64, LINUX_ARM64, OS_X_GO_1_5);

        // When we try to get the install package for an OS X version
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "Mac OS X", "x86_64", "10.11.12");

        // Then we should get the sole OS X package
        assertEquals("Got unexpected package", OS_X_GO_1_5, pkg);
    }

    @Test
    public void testLatestPackageForOsXVersionReturned() throws InstallationFailedException {
        // Given we have configured a release we want to install
        GolangRelease release = createReleaseInfo();

        // When we try to get the install package for a much newer OS X version
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "Mac OS X", "x86_64", "10.11.12");

        // Then we should get the newest package which is older than our version
        assertEquals("Got unexpected package", OS_X_10_8_64, pkg);
    }

    @Test
    public void testLatestPackageFor32BitOsXVersionReturned() throws InstallationFailedException {
        // Given we have configured a release we want to install
        GolangRelease release = createReleaseInfo();

        // When we try to get the install package for a much newer 32-bit OS X version
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "Mac OS X", "i386", "10.11.12");

        // Then we should get the newest package which is older than our version
        assertEquals("Got unexpected package", OS_X_10_8_32, pkg);
    }

    @Test
    public void testEarlierPackageForOsXVersionReturned() throws InstallationFailedException {
        // Given we have configured a release we want to install
        GolangRelease release = createReleaseInfo();

        // When we try to get the install package for an older, but supported OS X version
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "Mac OS X", "x86_64", "10.7");

        // Then we should get the newest package which is older than our version
        assertEquals("Got unexpected package", OS_X_10_6_64, pkg);
    }

    @Test
    public void testEarlierPackageFor32BitOsXVersionReturned() throws InstallationFailedException {
        // Given we have configured a release we want to install
        GolangRelease release = createReleaseInfo();

        // When we try to get the install package for an older, but supported 32-bit OS X version
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "Mac OS X", "i386", "10.7");

        // Then we should get the newest 32-bit package which is older than our version
        assertEquals("Got unexpected package", OS_X_10_6_32, pkg);
    }

    @Test
    public void testExactMatchPackageForOsXVersionReturned() throws InstallationFailedException {
        // Given we have configured a release we want to install
        GolangRelease release = createReleaseInfo();

        // When we try to get a install package which has an exact match on OS X version
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "Mac OS X", "x86_64", "10.6");

        // Then we should get the package which matches the given version exactly
        assertEquals("Got unexpected package", OS_X_10_6_64, pkg);
    }

    @Test(expected = InstallationFailedException.class)
    public void testUnsupportedOsXVersion() throws InstallationFailedException {
        // Given we have configured a release we want to install
        GolangRelease release = createReleaseInfo();

        // When we try to get a install package which has an exact match on OS X version
        GolangInstallable pkg = GolangInstaller.getInstallCandidate(release, "Mac OS X", "x86_64", "10.5");

        // Then an exception should be thrown
    }

    private static GolangRelease createReleaseInfo() {
        return createReleaseInfo(FREEBSD_32, FREEBSD_64, LINUX_32, LINUX_64, LINUX_ARM32, LINUX_ARM64,
                OS_X_10_6_32, OS_X_10_6_64, OS_X_10_8_32, OS_X_10_8_64);
    }

    private static GolangRelease createReleaseInfo(GolangInstallable... releases) {
        GolangRelease release = new GolangRelease();
        release.variants = releases;
        return release;
    }

    /**
     * @param os The OS value used in a Go archive filename.
     * @param arch The CPU architecture value used in a Go archive filename.
     */
    private static GolangInstallable createPackage(String os, String arch) {
        return createPackage(os, arch, null);
    }

    /**
     * @param os The OS value used in a Go archive filename.
     * @param arch The CPU architecture value used in a Go archive filename.
     * @param osxVersion Mac OS X version name.
     */
    private static GolangInstallable createPackage(String os, String arch, @Nullable String osxVersion) {
        GolangInstallable pkg = new GolangInstallable();
        pkg.os = os;
        pkg.arch = arch;
        pkg.osxversion = osxVersion;
        return pkg;
    }

}


src/main/java/org/jenkinsci/plugins/golang/GolangBuildWrapper.java
package org.jenkinsci.plugins.golang;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import hudson.CopyOnWrite;
import hudson.EnvVars;
import hudson.Extension;
import hudson.Launcher;
import hudson.model.AbstractBuild;
import hudson.model.AbstractProject;
import hudson.model.BuildListener;
import hudson.model.Computer;
import hudson.tasks.BuildWrapper;
import hudson.tasks.BuildWrapperDescriptor;
import org.kohsuke.stapler.DataBoundConstructor;

import java.io.IOException;
import java.util.Map;

public class GolangBuildWrapper extends BuildWrapper {

    private final String goVersion;

    @DataBoundConstructor
    public GolangBuildWrapper(String goVersion) {
        this.goVersion = goVersion;
    }

    @SuppressFBWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE")
    public Environment setUp(AbstractBuild build, Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
        GolangInstallation installation = getGoInstallation();
        if (installation != null) {
            EnvVars env = build.getEnvironment(listener);
            env.overrideAll(build.getBuildVariables());

            // Get the Go version for this node, installing it if necessary
            installation = installation.forNode(Computer.currentComputer().getNode(), listener).forEnvironment(env);
        }

        // Apply the GOROOT and go binaries to PATH
        final GolangInstallation install = installation;
        return new Environment() {
            @Override
            public void buildEnvVars(Map<String, String> env) {
                if (install != null) {
                    EnvVars envVars = new EnvVars();
                    install.buildEnvVars(envVars);
                    env.putAll(envVars);
                }
            }
        };
    }

    private GolangInstallation getGoInstallation() {
        for (GolangInstallation i : ((DescriptorImpl) getDescriptor()).getInstallations()) {
            if (i.getName().equals(goVersion)) {
                return i;
            }
        }
        return null;
    }

    @Extension
    public static class DescriptorImpl extends BuildWrapperDescriptor {

        @CopyOnWrite
        private volatile GolangInstallation[] installations = new GolangInstallation[0];

        public DescriptorImpl() {
            load();
        }

        @Override
        public String getDisplayName() {
            return Messages.SetUpGoTools();
        }

        @Override
        public boolean isApplicable(AbstractProject<?, ?> item) {
            return true;
        }

        @SuppressFBWarnings("EI_EXPOSE_REP")
        public GolangInstallation[] getInstallations() {
            return installations;
        }

        public void setInstallations(GolangInstallation... installations) {
            this.installations = installations;
            save();
        }

    }

}



src/main/java/org/jenkinsci/plugins/golang/GolangInstallation.java
package org.jenkinsci.plugins.golang;

import hudson.EnvVars;
import hudson.Extension;
import hudson.model.EnvironmentSpecific;
import hudson.model.Node;
import hudson.model.TaskListener;
import hudson.slaves.NodeSpecific;
import hudson.tools.ToolDescriptor;
import hudson.tools.ToolInstallation;
import hudson.tools.ToolInstaller;
import hudson.tools.ToolProperty;
import jenkins.model.Jenkins;
import org.jenkinsci.Symbol;
import org.kohsuke.stapler.DataBoundConstructor;

import java.io.File;
import java.io.IOException;
import java.util.Collections;
import java.util.List;

public class GolangInstallation extends ToolInstallation implements EnvironmentSpecific<GolangInstallation>,
        NodeSpecific<GolangInstallation> {

    @DataBoundConstructor
    public GolangInstallation(String name, String home, List<? extends ToolProperty<?>> properties) {
        super(name, home, properties);
    }

    @Override
    public void buildEnvVars(EnvVars env) {
        String root = getHome();
        if (root != null) {
            env.put("GOROOT", root);
            env.put("PATH+GOROOT_BIN", new File(root, "bin").toString());
        }
    }

    public GolangInstallation forEnvironment(EnvVars environment) {
        return new GolangInstallation(getName(), environment.expand(getHome()), getProperties().toList());
    }

    public GolangInstallation forNode(Node node, TaskListener log) throws IOException, InterruptedException {
        return new GolangInstallation(getName(), translateFor(node, log), getProperties().toList());
    }

    @Extension
    @Symbol("go")
    public static class DescriptorImpl extends ToolDescriptor<GolangInstallation> {

        @Override
        public String getDisplayName() {
            return "Go";
        }

        @Override
        public List<? extends ToolInstaller> getDefaultInstallers() {
            return Collections.singletonList(new GolangInstaller(null));
        }

        @Override
        public GolangInstallation[] getInstallations() {
            return Jenkins.getActiveInstance()
                    .getDescriptorByType(GolangBuildWrapper.DescriptorImpl.class)
                    .getInstallations();
        }

        @Override
        public void setInstallations(GolangInstallation... installations) {
            Jenkins.getActiveInstance()
                    .getDescriptorByType(GolangBuildWrapper.DescriptorImpl.class)
                    .setInstallations(installations);
        }

    }

}

src/main/java/org/jenkinsci/plugins/golang/GolangInstaller.java
package org.jenkinsci.plugins.golang;

import com.google.common.annotations.VisibleForTesting;
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import hudson.Extension;
import hudson.FilePath;
import hudson.model.DownloadService;
import hudson.model.Node;
import hudson.model.TaskListener;
import hudson.tools.DownloadFromUrlInstaller;
import hudson.tools.ToolInstallation;
import hudson.tools.ToolInstallerDescriptor;
import hudson.util.VersionNumber;
import jenkins.security.MasterToSlaveCallable;
import net.sf.json.JSONObject;
import org.kohsuke.stapler.DataBoundConstructor;

import java.io.IOException;
import java.net.URL;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/** Installs the Go programming language tools by downloading the archive for the detected OS/architecture combo. */
public class GolangInstaller extends DownloadFromUrlInstaller {

    @DataBoundConstructor
    public GolangInstaller(String id) {
        super(id);
    }

    // This is essentially the parent implementation, but we override it so we can pass Node into getInstallable()
    public FilePath performInstallation(ToolInstallation tool, Node node, TaskListener log) throws IOException,
            InterruptedException {
        FilePath expectedPath = preferredLocation(tool, node);

        Installable installable;
        try {
            installable = getInstallable(node);
        } catch (InstallationFailedException e) {
            throw new InstallationFailedException(Messages.CouldNotInstallGo(e.getMessage()));
        }

        if (installable == null) {
            log.getLogger().println(Messages.UnrecognisedReleaseId(id));
            return expectedPath;
        }

        if (isUpToDate(expectedPath, installable)) {
            return expectedPath;
        }

        String message = Messages.InstallingGoOnNode(installable.url, expectedPath, node.getDisplayName());
        if (expectedPath.installIfNecessaryFrom(new URL(installable.url), log, message)) {
            expectedPath.child(".timestamp").delete(); // we don't use the timestamp
            FilePath base = findPullUpDirectory(expectedPath);
            if (base != null && base != expectedPath)
                base.moveAllChildrenTo(expectedPath);
            // leave a record for the next up-to-date check
            expectedPath.child(".installedFrom").write(installable.url, "UTF-8");
        }

        return expectedPath;
    }

    @SuppressFBWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE")
    private Installable getInstallable(Node node) throws IOException, InterruptedException {
        // Get the Go release that we want to install
        GolangRelease release = getConfiguredRelease();
        if (release == null) {
            return null;
        }

        // Gather properties for the node to install on
        String[] properties = node.getChannel().call(new GetSystemProperties("os.name", "os.arch", "os.version"));

        // Get the best matching install candidate for this node
        return getInstallCandidate(release, properties[0], properties[1], properties[2]);
    }

    @VisibleForTesting
    static GolangInstallable getInstallCandidate(GolangRelease release, String osName, String osArch, String osVersion)
            throws InstallationFailedException {
        String platform = getPlatform(osName);
        String architecture = getArchitecture(osArch);

        // Sort and search for an appropriate variant
        List<GolangInstallable> variants = Arrays.asList(release.variants);
        Collections.sort(variants);
        for (GolangInstallable i : variants) {
            if (i.os.equals(platform) && i.arch.equals(architecture)) {
                if (i.osxversion == null) {
                    return i;
                }
                if (new VersionNumber(osVersion).compareTo(new VersionNumber(i.osxversion)) >= 0) {
                    return i;
                }
            }
        }

        String osWithVersion = osVersion == null ? osName : String.format("%s %s", osName, osVersion);
        throw new InstallationFailedException(Messages.NoInstallerForOs(release.name, osWithVersion, osArch));
    }

    private GolangRelease getConfiguredRelease() {
        List<GolangRelease> releases = ((DescriptorImpl) getDescriptor()).getInstallableReleases();
        if (releases == null) {
            return null;
        }

        for (GolangRelease r : releases) {
            if (r.id.equals(id)) {
                return r;
            }
        }
        return null;
    }

    @Extension
    public static final class DescriptorImpl extends ToolInstallerDescriptor<GolangInstaller> {
        public String getDisplayName() {
            return Messages.InstallFromWebsite();
        }

        // Used by config.groovy to show a human-readable list of releases
        @SuppressFBWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE")
        public List<GolangRelease> getInstallableReleases()  {
            return GolangReleaseList.all().get(GolangReleaseList.class).toList();
        }

        @Override
        public boolean isApplicable(Class<? extends ToolInstallation> toolType) {
            return toolType == GolangInstallation.class;
        }
    }

    @Extension
    public static final class GolangReleaseList extends DownloadService.Downloadable {

        // Public for JSON deserialisation
        public List<GolangRelease> releases;

        public GolangReleaseList() {
            super(GolangInstaller.class);
        }

        /** @return A list of available Go releases, obtained by parsing the JSON received from the update centre. */
        public List<GolangRelease> toList() {
            JSONObject root;
            try {
                root = getData();
                if (root == null) {
                    // JSON file has not yet been downloaded by Jenkins
                    return null;
                }
            } catch (IOException e) {
                // JSON parsing exception occurred
                return null;
            }

            Map<String, Class> classMap = new HashMap<String, Class>();
            classMap.put("releases", GolangRelease.class);
            classMap.put("variants", GolangInstallable.class);
            return ((GolangReleaseList) JSONObject.toBean(root, GolangReleaseList.class, classMap)).releases;
        }
    }

    // Needs to be public for JSON deserialisation
    public static class GolangRelease {
        public String id;
        public String name;
        public GolangInstallable[] variants;
    }

    // Needs to be public for JSON deserialisation
    public static class GolangInstallable extends Installable implements Comparable<GolangInstallable> {
        public String os;
        public String osxversion;
        public String arch;

        @SuppressFBWarnings("EQ_COMPARETO_USE_OBJECT_EQUALS")
        public int compareTo(GolangInstallable o) {
            // Sort by OS X version, descending
            if (osxversion != null && o.osxversion != null) {
                return new VersionNumber(o.osxversion).compareTo(new VersionNumber(osxversion));
            }
            // Otherwise we don't really care; sort by OS name
            return os.compareTo(o.os);
        }

        @Override
        public String toString() {
            return String.format("GolangInstallable[os=%s, arch=%s, version=%s]", os, arch, osxversion);
        }
    }

    /** @return The OS value used in a Go archive filename, for the given {@code os.name} value. */
    private static String getPlatform(String os) throws InstallationFailedException {
        String value = os.toLowerCase(Locale.ENGLISH);
        if (value.contains("freebsd")) {
            return "freebsd";
        }
        if (value.contains("linux")) {
            return "linux";
        }
        if (value.contains("os x")) {
            return "darwin";
        }
        if (value.contains("windows")) {
            return "windows";
        }
        throw new InstallationFailedException(Messages.UnsupportedOs(os));
    }

    /** @return The CPU architecture value used in a Go archive filename, for the given {@code os.arch} value. */
    private static String getArchitecture(String arch) throws InstallationFailedException {
        String value = arch.toLowerCase(Locale.ENGLISH);
        if (value.contains("amd64") || value.contains("86_64")) {
            return "amd64";
        }
        // There seems to be no canonical documentation on what the `os.arch` values are for different ARM CPU types,
        // and presumably there's no real guarantee of consistency between JVM vendors anyway.
        // It _seems_ to be `aarch64` and `arm` for 64- and 32-bit, but let's also have some reasonable fallbacks…
        if (value.contains("aarch64") || value.contains("arm64")) {
            return "arm64";
        }
        if (value.contains("aarch32") || value.contains("arm")) {
            return "armv6l";
        }
        if (value.contains("86")) {
            return "386";
        }
        if (value.contains("s390x")) {
            return "s390x";
        }
        throw new InstallationFailedException(Messages.UnsupportedCpuArch(arch));
    }

    /** Returns the values of the given Java system properties. */
    private static class GetSystemProperties extends MasterToSlaveCallable<String[], InterruptedException> {
        private static final long serialVersionUID = 1L;

        private final String[] properties;

        GetSystemProperties(String... properties) {
            this.properties = properties;
        }

        public String[] call() {
            String[] values = new String[properties.length];
            for (int i = 0; i < properties.length; i++) {
                values[i] = System.getProperty(properties[i]);
            }
            return values;
        }
    }

    // Extend IOException so we can throw and stop the build if installation fails
    static class InstallationFailedException extends IOException {
        InstallationFailedException(String message) {
            super(message);
        }
    }

}
